[SystemZ] Implement .machine (push|pop) directives

dominik-steenken opened the pull request:
The `.machine push` and `.machine pop` directives were missing from the SystemZ Backend Asm Parser. This PR adds them, and expands the corresponding test to test proper operation.

Internally, this is modeled as a simple stack implemented on a `SmallVector<StringRef>`.

dominik-steenken committed:
[SystemZ] Implement .machine (push|pop) directives

The `.machine push` and `.machine pop` directives were missing from
the SystemZ Backend Asm Parser. This commit adds them, and expands the
corresponding test to test proper operation.

dominik-steenken commented:
@uweigand FYI

uweigand commented:
> This implementation makes the assumption that, without any `.machine` directive, the "current" machine is undefined, rather than trying to derive it based on the host machine. Thus, in this implementation, any `.machine push` _must_ be preceded by at least one `.machine <cpu>`. Otherwise, a parsing error is recognized.

I don't think this matches GAS behavior.   Why is this necessary?   Can't we just push the feature bits instead of the name onto the stack?  There's the default set of feature bits that's already detected (and used before any explicit machine directive); a first push would just push this; subsequent pops would restore whatever bits are set without having to re-compute them from a name.

dominik-steenken commented on llvm/lib/Target/SystemZ/AsmParser/SystemZAsmParser.cpp:
```diff
@@ -410,6 +411,18 @@ class SystemZAsmParser : public MCTargetAsmParser {
+  // it always reflects the current FeatureBitset used to determine which
+  // machine features are available, i.e. either the default or the result
+  // of the most reecent `.machine` directive.
+  FeatureBitset CurrentFeatures;
```
Good point, i have removed the copy and use `getAvailableFeatures()` instead.

dominik-steenken commented on llvm/lib/Target/SystemZ/AsmParser/SystemZAsmParser.cpp:
```diff
@@ -1382,17 +1395,35 @@ bool SystemZAsmParser::parseDirectiveMachine(SMLoc L) {
+      return TokError("pop without corresponding push in '.machine' directive");
+    CurrentFeatures = MachineStack.back();
+    MachineStack.pop_back();
+    setAvailableFeatures(CurrentFeatures);
```
Thanks, i was a bit confused as to what `ComputeAvailableFeatures` did since it takes and produces a `FeatureBitset`. Since i took your advice above, i think this is now addressed as well.

dominik-steenken commented:
> > This implementation makes the assumption that, without any `.machine` directive, the "current" machine is undefined, rather than trying to derive it based on the host machine. Thus, in this implementation, any `.machine push` _must_ be preceded by at least one `.machine <cpu>`. Otherwise, a parsing error is recognized.
> 
> I don't think this matches GAS behavior. Why is this necessary? Can't we just push the feature bits instead of the name onto the stack? There's the default set of feature bits that's already detected (and used before any explicit machine directive); a first push would just push this; subsequent pops would restore whatever bits are set without having to re-compute them from a name.

Thanks, that was very helpful. I was somewhat unhappy with this solution as well. It originated from me having difficulties obtaining a string description of the CPU from the `SubTargetInfo` - it seemed no matter where i sourced this from, it would always be an empty string. But as you point out, it is also possible to just store the bits and restore them once the `.machine` stack is empty. I have updated the commit to provide an implementation that does this. `.machine push` without a prior `.machine <cpu>` is thus now legal and i have modified the tests accordingly. I have also tried to incorporate your additional feedback. I am now waiting for the CI to complete.

uweigand commented on llvm/lib/Target/SystemZ/AsmParser/SystemZAsmParser.cpp:
```diff
@@ -1382,16 +1388,32 @@ bool SystemZAsmParser::parseDirectiveMachine(SMLoc L) {
+  // Do error check before lexing the next token so that TokError
+  // points at the correct token.
+  if ((Id == "pop") && (MachineStack.empty()))
+    return TokError("pop without corresponding push in '.machine' directive");
```
Hmm.  I do think it would be better to have the syntax check before the semantics check.  E.g. if you do a `.machine pop z15` or something, you should preferably see the error that there's an unexpected token at the end, over the error that there was no corresponding push.

To get the location correct, I think it would be better to use an `Error` with an explicit location (e.g. by calling `Parser.getTok().getLoc()` at the place where you do the `getIdentifier` call, and then using that location with `Error` further down).

dominik-steenken replied:
makes sense to me, i have reincorporated the check for the semantic error and added an additional line to the error test to check that the `.machine pop <cpu>` situation you outlined does in fact point at the syntax issue and not the semantic one.

dominik-steenken replied:
actually hang on i used a push instead of a pop...

dominik-steenken replied:
and it does report the semantic issue first :/
GIve me a second please

dominik-steenken replied:
resolved. It works now, at least in my local testing.

uweigand reviewed (approved):
LGTM now assuming CI passes.

dominik-steenken commented:
From what i can tell, this `lldb` check fail exists on `main` as well. Not sure what to do about this, for now i'll just rebase.

uweigand commented:
Thanks.  The Windows failure looks unrelated, so I think this is good to go now.
